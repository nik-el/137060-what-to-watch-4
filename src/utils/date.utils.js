const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const getFormattedDate = (dateString) => {
  if (!dateString) {
    return ``;
  }

  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
