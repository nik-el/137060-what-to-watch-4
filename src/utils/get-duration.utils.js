const HOUR_MIN = 60;
const HOUR_SEC = 3600;

export const geFormatDuration = (time) => {
  const hours = Math.floor(time / HOUR_MIN);
  const min = time % HOUR_MIN;

  return `${hours}h ${min}m`;
};

export const geVideoDuration = (time) => {
  const secNum = parseInt(time, 10);
  let hours = Math.floor(secNum / HOUR_SEC);
  let minutes = Math.floor((secNum - (hours * HOUR_SEC)) / HOUR_MIN);
  let seconds = secNum - (hours * HOUR_SEC) - (minutes * HOUR_MIN);

  if (hours < 10) {
    hours = `0` + hours;
  }
  if (minutes < 10) {
    minutes = `0` + minutes;
  }
  if (seconds < 10) {
    seconds = `0` + seconds;
  }
  return hours + `:` + minutes + `:` + seconds;
};
