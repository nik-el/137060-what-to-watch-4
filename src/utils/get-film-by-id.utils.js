export const getFilmById = (films, id) => {
  if (!Array.isArray(films) || !id) {
    return null;
  }

  return films.find((item) => item.id === id);
};
