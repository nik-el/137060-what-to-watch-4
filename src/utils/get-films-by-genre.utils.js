export const getFilmsByGenre = (films, genre) => {
  if (!Array.isArray(films)) {
    return [];
  }
  if (!genre) {
    return films;
  }

  return films.filter((item) => item.genre === genre);
};
