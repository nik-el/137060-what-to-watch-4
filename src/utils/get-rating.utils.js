export const getRating = (rating) => {
  switch (true) {
    case (rating < 3):
      return `Bad`;
    case (rating >= 3 && rating < 5):
      return `Normal`;
    case (rating >= 5 && rating < 8):
      return `Good`;
    case (rating >= 8 && rating < 10):
      return `Very good`;
    case (rating === 10):
      return `Awesome`;
    default:
      return `No rating`;
  }
};
