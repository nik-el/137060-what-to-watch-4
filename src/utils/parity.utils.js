const getOddArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }
  return array.filter((item, index) => {
    return index % 2;
  });

};
const getEvenArray = (array) => {
  if (!Array.isArray(array)) {
    return [];
  }

  return array.filter((item, index) => {
    return !(index % 2);
  });
};

export {
  getOddArray,
  getEvenArray
};
