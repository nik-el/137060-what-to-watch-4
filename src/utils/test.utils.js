const testFilm = {
  title: `Test title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testFilm);

const testEditedFilm = {
  title: `Test edit title 1`,
  genre: `test genre 1`,
  releaseYear: 2000,
  thumbnail: `https://picsum.photos/id/1`,
  bgImage: `https://picsum.photos/id/1`,
  bgColor: `#FFF`,
  poster: `https://picsum.photos/id/1`,
  description: `Test description`,
  rating: 5.5,
  ratingCount: 100,
  director: `Test director`,
  starring: [`Test Star 1`, `Test Star 2`],
  id: `1`,
  preview: `https://picsum.photos/id/1`,
  runTime: 100,
};
Object.freeze(testEditedFilm);

const testFilms = [
  {
    title: `Test title 1`,
    genre: `test genre 1`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/1`,
    bgImage: `https://picsum.photos/id/1`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/1`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `1`,
    preview: `https://picsum.photos/id/1`,
    runTime: 100,
  },
  {
    title: `Test title 2`,
    genre: `test genre 2`,
    releaseYear: 2000,
    thumbnail: `https://picsum.photos/id/2`,
    bgImage: `https://picsum.photos/id/2`,
    bgColor: `#FFF`,
    poster: `https://picsum.photos/id/2`,
    description: `Test description`,
    rating: 5.5,
    ratingCount: 100,
    director: `Test director`,
    starring: [`Test Star 1`, `Test Star 2`],
    id: `2`,
    preview: `https://picsum.photos/id/2`,
    runTime: 100,
  },
];
Object.freeze(testFilms);

const testGenres = [`test genre 1`, `test genre 2`];
Object.freeze(testGenres);

const testComments = [{
  id: 1,
  user: {
    id: 4,
    name: `Test  name`
  },
  rating: 1.1,
  comment: `test comment`,
  date: `2019-05-08T14:13:56.569Z`
}];
Object.freeze(testComments);

const testStore = {
  DATA: {
    films: testFilms,
    genres: testGenres,
    loadingFilmsData: false,
    restApi: `test_api_url`
  },
  VIEW: {
    currentGenre: testGenres[0],
    feedLimit: 10,
  },
  USER: {
    authorizationStatus: `TEST_AUTH`,
    user: {
      "id": 1,
      "email": `test@test.com`,
      "name": `test`,
      "avatarUrl": `test/1.png`
    }
  },
  REVIEW: {
    loadingCommentsData: false,
    loadingAddingComment: false,
    comments: testComments
  },
  FAVORITE: {
    favorites: testFilms,
    loadingFavoritesData: false,
    loadingSetFavorite: false
  }
};
Object.freeze(testStore);

export {testFilm, testFilms, testEditedFilm, testGenres, testComments, testStore};
