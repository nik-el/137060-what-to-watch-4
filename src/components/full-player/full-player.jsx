import React, {useRef, useState} from 'react';
import PropTypes from 'prop-types';
import {useParams} from "react-router-dom";
import {geVideoDuration} from "../../utils/get-duration.utils";
import history from './../../history';

export const FullPlayer = React.memo(function FullPlayer({currentFilms}) {
  const videoEl = useRef(null);
  const [timeElapsed, setTimeElapsed] = useState(`00:00:00`);
  const [progressPercent, setProgressPercent] = useState(`0`);
  const [isPlaying, setPlayingStatus] = useState(false);

  const {id} = useParams();
  const film = currentFilms.find((item) => item.id === id);

  const handleVideoTimeUpdate = () => {
    const time = geVideoDuration(videoEl.current.duration - videoEl.current.currentTime);

    const progress = 100 * videoEl.current.currentTime / videoEl.current.duration;
    setProgressPercent(progress.toFixed(2));
    setTimeElapsed(time);
  };

  const handleVideoPlayStatus = () => {
    const video = videoEl.current;
    setPlayingStatus(!video.paused);
  };

  const handlePlayClick = () => {
    const video = videoEl.current;
    if (isPlaying) {
      video.pause();
    } else {
      video.play();
    }

    setPlayingStatus(!video.paused);
  };

  const handleFullScreenClick = () => {
    const video = videoEl.current;
    video.requestFullscreen();
  };

  const handleExitClick = () => {
    const video = videoEl.current;
    video.pause();
    history.goBack();
  };

  return <div className="player">
    <video
      autoPlay
      src={film.video}
      className="player__video"
      poster={film.bgImage}
      ref={videoEl}
      onTimeUpdate={handleVideoTimeUpdate}
      onPlay={handleVideoPlayStatus}
    >
    </video>

    <button
      type="button"
      className="player__exit"
      onClick={handleExitClick}
    >
      Exit
    </button>

    <div className="player__controls">
      <div className="player__controls-row">
        <div className="player__time">
          <progress className="player__progress" value={progressPercent} max="100"/>
          <div
            className="player__toggler"
            style={{
              left: progressPercent + `%`
            }}
          >Toggler</div>
        </div>
        <div className="player__time-value">{ timeElapsed }</div>
      </div>

      <div className="player__controls-row">
        <button
          type="button"
          className="player__play"
          onClick={handlePlayClick}
        >
          <svg viewBox="0 0 19 19" width="19" height="19">
            <use xlinkHref={isPlaying ? `#pause` : `#play-s`}/>
          </svg>
          <span>Play</span>
        </button>
        <div className="player__name">{film.title}</div>

        <button
          type="button"
          className="player__full-screen"
          onClick={handleFullScreenClick}
        >
          <svg viewBox="0 0 27 27" width="27" height="27">
            <use xlinkHref="#full-screen"/>
          </svg>
          <span>Full screen</span>
        </button>
      </div>
    </div>
  </div>;

});

FullPlayer.propTypes = {
  currentFilms: PropTypes.arrayOf(PropTypes.object),
};
