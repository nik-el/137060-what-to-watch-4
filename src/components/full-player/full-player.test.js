import React from 'react';
import renderer from 'react-test-renderer';
import {FullPlayer} from './full-player';
import {testFilms} from '../../utils/test.utils';
import {MemoryRouter, Route} from "react-router-dom";

const currentTestFilmId = testFilms[0].id;

it(`FullPlayer renders correctly`, () => {
  const tree = renderer
    .create(
        <MemoryRouter initialEntries={[`player/${currentTestFilmId}`]}>
          <Route path={`player/:id`}>
            <FullPlayer
              currentFilms={testFilms}
            />
          </Route>
        </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

