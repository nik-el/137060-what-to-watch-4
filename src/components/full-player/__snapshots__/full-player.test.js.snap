// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullPlayer renders correctly 1`] = `null`;
